### PR TITLE
Introduce ProviderOption and small cleanup

### PR DESCRIPTION
**Pull request by DavidSGK**
# Details
Some small changes
- Introduce `ProviderOption` interface and update both API and sidecar provider initializations to take options
    - Add `URLOption` as an option to override API URL for API provider mode
- Remove API key requirement from sidecar provider
- Mark some variables as private

### CODE REVIEW

File: client/api_provider.go [hunk 1]
```diff
@@ -281,3 +286,17 @@ func (a *apiProvider) GetJSONFeature(ctx context.Context, key string, namespace
...
+}
+
+// Used to override the default Lekko backend API URL.
+// Only applicable to API provider.
+type URLOption struct {
+	URL string
+}
+
+func (uo *URLOption) apply(provider *apiProvider) {
+	provider.lekkoClient = clientv1beta1connect.NewConfigurationServiceClient(http.DefaultClient, uo.URL)
```
Comment (passichenko) [inline client/api_provider.go:301]: Allowing custom options to directly change internal sounds a bit hack-y. What about having "options" struct that custom options can change and then using it during initialization?

**Review by passichenko — COMMENTED** [review]
Neat, is it go-way to do optional arguments :) ? LG, but I have one question (see comment on the diff).

File: client/api_provider.go (hunk 1 above)
Comment (DavidSGK) [reply client/api_provider.go:301]: If I understand correctly, I think it's fine and behavior I've seen similarly in other libraries - the pattern is that options that implement the interface do have direct access to modify the internals of what they're options for.

**Comment by DavidSGK** [comment]
> Neat, is it go-way to do optional arguments :) ? LG, but I have one question (see comment on the diff).

I think so, yeah it's kind of a cool way of (hopefully) doing things in a canonically Go way. Since Go doesn't have optional parameters, default values, or overloading, this kind of option handling is a clean way of allowing users to not have to pass in `nil` for empty args or something like that.

**Review by konradjniemiec — APPROVED** [review]

**Pull request merged**